src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveReviewsService.kt

```diff
@@ -42,44 +48,66 @@ class IncentiveReviewsService(
+
     val deferredPositiveCaseNotesInLast3Months = async { getCaseNoteUsage("POS", "IEP_ENC", prisonerNumbers) }
     val deferredNegativeCaseNotesInLast3Months = async { getCaseNoteUsage("NEG", "IEP_WARN", prisonerNumbers) }
 
-    val bookingIds = offenders.content.map(OffenderSearchPrisoner::bookingId)
     val nextReviewDates = getNextReviewDatesForOffenders(bookingIds)
-
-    val bookingIdsMissingIepReviews = bookingIds.toSet() subtract nextReviewDates.keys
+    val bookingIdsMissingIepReviews = bookingIdSet subtract nextReviewDates.keys
     if (bookingIdsMissingIepReviews.isNotEmpty()) {
-      throw ListOfDataNotFoundException(bookingIdsMissingIepReviews)
+      throw ListOfDataNotFoundException("next review dates", bookingIdsMissingIepReviews)
     }
+
     val overdueCount = nextReviewDates.map { r -> r.value.nextReviewDate }.count { it.isBefore(LocalDate.now(clock)) }
 
+    val incentiveLevels = deferredIncentiveLevels.await()
+    val bookingIdsMissingIncentiveLevel = bookingIdSet subtract incentiveLevels.keys
+    if (bookingIdsMissingIncentiveLevel.isNotEmpty()) {
+      throw ListOfDataNotFoundException("incentive levels", bookingIdsMissingIncentiveLevel)
+    }
+
     val positiveCaseNotesInLast3Months = deferredPositiveCaseNotesInLast3Months.await()
     val negativeCaseNotesInLast3Months = deferredNegativeCaseNotesInLast3Months.await()
 
     val locationDescription = deferredLocationDescription.await()
 
-    IncentiveReviewResponse(
-      reviewCount = offenders.totalElements,
-      overdueCount = overdueCount,
-      locationDescription = locationDescription,
-      reviews = offenders.content.map {
+    val reviews = offenders.content
+      .map {
         IncentiveReview(
           prisonerNumber = it.prisonerNumber,
           bookingId = it.bookingId,
           firstName = WordUtils.capitalizeFully(it.firstName),
           lastName = WordUtils.capitalizeFully(it.lastName),
+          levelCode = incentiveLevels[it.bookingId]!!.iepCode,
           positiveBehaviours = positiveCaseNotesInLast3Months[it.prisonerNumber]?.totalCaseNotes ?: 0,
           negativeBehaviours = negativeCaseNotesInLast3Months[it.prisonerNumber]?.totalCaseNotes ?: 0,
           acctOpenStatus = it.acctOpen,
           nextReviewDate = nextReviewDates[it.bookingId]!!.nextReviewDate,
         )
-      }.sortedBy { it.nextReviewDate }
+      }
+      .filter { it.levelCode == levelCode }
+      .sortedBy { it.nextReviewDate }
+
+    IncentiveReviewResponse(
+      locationDescription = locationDescription,
+      overdueCount = overdueCount,
+      reviewCount = reviews.size,
+      reviews = reviews,
     )
   }
 
+  private suspend fun getIncentiveLevelsForOffenders(bookingIds: List<Long>): Map<Long, PrisonerIepLevel> =
+    prisonerIepLevelRepository.findAllByBookingIdInAndCurrentIsTrueOrderByReviewTimeDesc(bookingIds)
+      .fold(mutableMapOf()) { map, prisonerIepLevel ->
```

Comment:
I think you could also achieve this with `.toList().associateBy(PrisonerIepLevel::bookingId)` which is maybe slightly simpler?

Reply:
> i think this function will change as the "next review date" stuff gets iterated 

good point. let's see what comes out of that and agreed may be able to refactor

Reply:
i've now put in something i think is more ergonomic… is it better?